src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp

```diff
@@ -1464,23 +1527,57 @@ INSStaggeredHierarchyIntegrator::setupSolverVectors(const Pointer<SAMRAIVectorRe
     // Account for body forcing terms.
     if (d_F_fcn)
     {
-        d_F_fcn->setDataOnPatchHierarchy(d_F_scratch_idx, d_F_var, d_hierarchy, half_time);
+        double data_time;
+        if (is_bdf_time_stepping_type(d_viscous_time_stepping_type))
+        {
+            data_time = new_time;
+        }
+        else
+        {
+            data_time = half_time;
+        }
+        d_F_fcn->setDataOnPatchHierarchy(d_F_scratch_idx, d_F_var, d_hierarchy, data_time);
```

Comment:
This might be a problem for body forcing terms that depend on advected quantities, for example I don't know if the complex fluid solver is set up to correctly handle evaluations at `new_time`.

I have switched over to using AB2 for body forces.

Reply:
This change is actually the primary reason that I am doing this --- we want to be able to use the end-time step forcing for IIM.

Reply:
Yes, but we need a better way to inform force functions that the approximation should be using data at `t^{n+1}` instead of `t^{n+1/2}`. For example, this always uses `t^{n+1/2}` regardless of time stepping scheme, which for BDF2 will be first order.
https://github.com/IBAMR/IBAMR/blob/2b40ae33badfd03f2759c8999705cf91060da548/examples/navier_stokes/ex5/BoussinesqForcing.cpp#L76-L86

Reply:
Well, here the implementation (probably by me) is ignoring `data_time`....!

Reply:
There is a hook in the API that callers can use to inform when things should be evaluated. If the called function cannot evaluate at that time, it should just emit an error message and abort.

Reply:
I agree, but there is not a great API to figure out which time point `data_time` corresponds to. But this is an issue with the use of `CartGridFunction`, not the time stepping scheme.

Reply:
I've tried to capture this issue and a possible approach in #1709.

Reply:
Also #1710. It would be nice to be able to handle time extrapolation in a more uniform manner, so that (e.g.) we could easily allow for either evaluation at the end of the time step or extrapolation to the end of the time step in places like this.